ethereum/sidecar/server.go

```diff
@@ -434,6 +581,387 @@ func (s *Server) startGRPCServer(
+			"failed to fetch assets unlocked sequence tip: [%w]",
+			err,
+		)
+	}
+
+	// The sequence tip is zero, meaning no events have been made.
+	if sequenceTip.IsZero() {
+		return []bridgetypes.AssetsUnlockedEvent{}, nil
+	}
+
+	// Cut-off block time in UNIX seconds.
+	cutOffBlockTime := uint32(mezotime.Now().Add(-s.assetsUnlockedLookBackPeriod).Unix()) //nolint:gosec
+	recentEvents := []bridgetypes.AssetsUnlockedEvent{}
+
+	// Walk backwards from the current sequence tip in windows of at most
+	// `assetsUnlockedBatchSize`.
+outer:
+	for sequenceTip.IsPositive() {
+		// Sequence end is exclusive. We must add `1`.
+		seqEnd := sequenceTip.AddRaw(1)
+
+		// Sequence start is inclusive. It cannot be lower than `1`.
+		seqStart := seqEnd.SubRaw(int64(s.assetsUnlockedBatchSize))
+		if seqStart.LT(sdkmath.OneInt()) {
+			seqStart = sdkmath.OneInt()
+		}
+
+		events, err := s.assetsUnlockedEndpoint.GetAssetsUnlockedEvents(
+			ctx,
+			seqStart,
+			seqEnd,
+		)
+		if err != nil {
+			return nil, fmt.Errorf(
+				"failed to fetch AssetsUnlocked events [%s, %s): %w",
+				seqStart.String(),
+				seqEnd.String(),
+				err,
+			)
+		}
+
+		for i := len(events) - 1; i >= 0; i-- {
+			event := events[i]
+			if event.BlockTime < cutOffBlockTime {
+				// We found an event older than cut-off time.
+				break outer
+			}
+			recentEvents = append(recentEvents, event)
+		}
+
+		// If the length of fetched events was smaller than the batch size,
+		// there are no more events to fetch.
+		if len(events) < s.assetsUnlockedBatchSize {
+			break
+		}
+
+		sequenceTip = seqStart.SubRaw(1)
+	}
+
+	// Ensure events are sorted in ascending order of their unlock sequences.
+	sort.Slice(recentEvents, func(i, j int) bool {
+		return recentEvents[i].UnlockSequence.LT(recentEvents[j].UnlockSequence)
+	})
+
+	return recentEvents, nil
+}
+
+// findUnconfirmedAssetsUnlockedEvents finds unconfirmed `AssetsUnlocked` events
+// from the passed input events. An event is considered confirmed if there
+// was an `AssetsUnlockConfirmed` event emitted for it on Ethereum within
+// finalized range of blocks.
+func (s *Server) findUnconfirmedAssetsUnlockedEvents(
+	ctx context.Context,
+	inputEvents []bridgetypes.AssetsUnlockedEvent,
+) ([]bridgetypes.AssetsUnlockedEvent, error) {
+	// If there are no input events, return early with an empty list.
+	if len(inputEvents) == 0 {
+		return []bridgetypes.AssetsUnlockedEvent{}, nil
+	}
+
+	currentBlock, err := s.chain.BlockCounter().CurrentBlock()
+	if err != nil {
+		return nil, fmt.Errorf(
+			"failed to get current block on Ethereum chain [%w]",
+			err,
+		)
+	}
+
+	// The search range can be limited to avoid excessive chain data usage.
+	// When defining the start block we must make sure we cover the entire range
+	// in which the input events could have been confirmed on Ethereum.
+	startBlock := uint64(0)
+	if currentBlock > assetsUnlockConfirmedLookBackBlocks {
+		startBlock = currentBlock - assetsUnlockConfirmedLookBackBlocks
```

Comment:
Shouldn't we count back from finalized block? What's the point of doing so from current block?

Reply:
Done in 7ec4a2fca2268a922f4759ada90a2ab1144acd2f.
My thinking was: search for `AssetsUnlockConfirmed` events over the last (i.e. since the current tip) ~1 month worth of blocks. But since the difference between the current and finalized blocks is at most 20 minutes, it doesn't matter.